takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/AbstractCompileMojo.java

```diff
@@ -448,63 +448,54 @@ public void execute() throws MojoExecutionException, MojoFailureException {
     }
   }
 
-  private void checkDependencyReferences(Set<String> referencedEntries) throws MojoExecutionException, IOException {
+  private void checkDependencyReferences(Set<File> referencedEntries) throws MojoExecutionException, IOException {
     Set<Artifact> referencedArtifacts = new LinkedHashSet<>();
     Set<Artifact> undeclaredArtifacts = new LinkedHashSet<>();
     Set<Artifact> unusedArtifacts = new LinkedHashSet<>();
-    List<File> processorpath = getProcessorpath();
+    List<File> processorPath = getProcessorpath();
+
+    // Include processor path in referenced entries
+    Set<File> allReferencedEntries = new LinkedHashSet<>();
+    allReferencedEntries.addAll(referencedEntries);
+    if (processorPath != null) {
+      allReferencedEntries.addAll(processorPath);
+    }
 
     // Find the equivalent artifact equivalent for each referenced entry
-    for (String entry : referencedEntries) {
-      for (Artifact artifact : getClasspathArtifacts()) {
-        if (entry.equals(artifact.getFile().getCanonicalPath())) {
-          referencedArtifacts.add(artifact);
-        }
+    for (File entry : allReferencedEntries) {
+      Artifact artifact = getClasspath().get(entry);
+      if (artifact != null) {
+        referencedArtifacts.add(artifact);
       }
     }
 
     if (directDependencies != null) {
       // Check each direct dependency for existence in referenced artifacts
       for (Artifact dependency : directDependencies) {
         if (!referencedArtifacts.contains(dependency)) {
-          // Check if artifact is used for APT
-          if (processorpath == null || !processorpath.contains(dependency.getFile())) {
-            unusedArtifacts.add(dependency);
-          }
+          unusedArtifacts.add(dependency);
         }
       }
 
       // Check all referenced artifacts to see if they're declared as a direct dependency
       for (Artifact artifact : referencedArtifacts) {
-        boolean found = false;
-        Iterator<Artifact> it = directDependencies.iterator();
-        while (it.hasNext()) {
-          if (artifact.equals(it.next())) {
-            found = true;
-            break;
-          }
-        }
-        if (!found) {
+        if (!directDependencies.contains(artifact)) {
           undeclaredArtifacts.add(artifact);
         }
       }
     }
 
-    // Log or fail the build for any referenced artifacts that are not declared as a direct dependency
+    // Fail the build for any referenced artifacts that are not declared as a direct dependency
     if (!undeclaredArtifacts.isEmpty()) {
       if (unusedDeclaredDependency == AccessRulesViolation.error) {
```

Comment:
Looking at this again. Would it make sense to have a separate `AccessRulesViolation` for "undeclared" vs "unused" artifacts? To fail only on undeclared or unused rather than failing on any combination of the two. Although, I guess technically speaking, "undeclared" is almost the same thing as `transitiveDependencyReference`.

Reply:
You are right, I missed it.

`checkDependencyReferences` should really be `checkUnusedDependencies` and only check for unused dependencies declared in project pom.xml (or inherited from the parent).

Existing `transitiveDependencyReference` policy already covers references to transitive/undeclared dependencies. No need duplicate functionality. I'll update review in the code too.